### PR TITLE
make it possible to query by artifact provides

### PR DESCRIPTION
**Pull request by kjaskiewiczz**
Ticket: [Men 6107](https://tracker.mender.io/browse/MEN-6107)

### CODE REVIEW

**Comment by kjaskiewiczz** [comment]
@mender-test-bot start pipeline --fast

File: model/image.go [hunk 1]
```diff
@@ -31,8 +32,19 @@ import (
...
+	runeDollar         = '\uFF04'
+	runeDot            = '\uFF0E'
 )
 
+var (
+	StorageKeyImageProvidesRootFSChecksum = "meta_artifact.provides." +
+		GetProvidesKeyReplacer().Replace("rootfs-image.checksum")
+	StorageKeyImageProvidesRootFSVersion = "meta_artifact.provides." +
+		GetProvidesKeyReplacer().Replace("rootfs-image.version")
+)
```
Comment (tranchitella) [inline model/image.go:44]: Not tested, but you get the idea... why initializing the replacer each time you use it?

```suggestion
var (
        providesKeyReplacer = GetProvidesKeyReplacer()
	StorageKeyImageProvidesRootFSChecksum = "meta_artifact.provides." +
		providesKeyReplacer.Replace("rootfs-image.checksum")
	StorageKeyImageProvidesRootFSVersion = "meta_artifact.provides." +
		providesKeyReplacer.Replace("rootfs-image.version")
)
```

File: model/image.go [hunk 2]
```diff
@@ -267,3 +279,18 @@ func (msg MultipartGenerateImageMsg) Validate() error {
...
+
+func GetProvidesKeyReplacer() *strings.Replacer {
+	return strings.NewReplacer(".", string(runeDot), "$", string(runeDollar))
+}
+
+// MarshalBSONValue marshals the Provides to a mongo-compatible
+// document.
+func (p Provides) MarshalBSONValue() (bsontype.Type, []byte, error) {
+	attrs := make(bson.M, len(p))
+	replacer := GetProvidesKeyReplacer()
```
Comment (tranchitella) [inline model/image.go:291]:
```suggestion
```

File: model/image.go [hunk 3]
```diff
@@ -267,3 +279,18 @@ func (msg MultipartGenerateImageMsg) Validate() error {
...
+	for k, v := range p {
+		attrs[replacer.Replace(k)] = v
```
Comment (tranchitella) [inline model/image.go:293]:
```suggestion
		attrs[providesKeyReplacer.Replace(k)] = v
```

File: model/image.go (hunk 1 above)
Comment (kjaskiewiczz) [reply model/image.go]: done

File: model/image.go (hunk 2 above)
Comment (kjaskiewiczz) [reply model/image.go]: done

File: model/image.go (hunk 3 above)
Comment (kjaskiewiczz) [reply model/image.go]: done (the last commit)

File: store/mongo/datastore_mongo.go [hunk 4]
```diff
@@ -325,6 +328,20 @@ var (
...
+	IndexArtifactProvidesRootfsVersionAndChecksum = mongo.IndexModel{
+		Keys: bson.D{
+			{Key: model.StorageKeyImageProvidesRootFSVersion,
+				Value: 1},
+			{Key: model.StorageKeyImageProvidesRootFSChecksum,
+				Value: 1},
+		},
+		Options: &mopts.IndexOptions{
+			Background: &_false,
+			Name:       &IndexArtifactProvidesRootfsVersionAndChecksumName,
```
Comment (alfrunes) [inline store/mongo/datastore_mongo.go:342]: This index needs to be sparse or use a `PartialFilterExpression` to cover documents without Provides.

**Comment by kjaskiewiczz** [comment]
> The index is not compatible with the new schema.

not sure I get it; I would appreciate an explanation for dummies ;)

> General question: Since we're migrating all documents, maybe we should consider changing the schema for provides to hold embedded documents of key/value pairs? It might make the index a bit fatter, but it adds the ability to query arbitrary provides values if there should be a need.

are embedded documents in depends/provides supported by the mender-artifact?

**Comment by kjaskiewiczz** [comment]
> I think we impose that provides/depends needs to be a flat string mapping. In that case, my question is if we should encode these key/value pairs as embedded documents in MongoDB on the form:
> 
> ```go
> {"provides": [{"key": "<provides1>", "value": "value1"},{"key": "<provides2>", "value": "<value2>"}, ...]}
> ```

currently I'm creating index with rootfs-image.checksum and rootfs-image.versin only; with embedded documents I'd have to index all provides, right (index for provides.key)? 
the index will potentially be quite big, but it would solve [this issue](https://tracker.mender.io/browse/MEN-6063?focusedCommentId=128669&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-128669)

File: store/mongo/datastore_mongo.go (hunk 4 above)
Comment (kjaskiewiczz) [reply store/mongo/datastore_mongo.go:342]: it's a regular index (not unique), so, as you pointed out, there is no need for PartialFilterExpression

**Review by tranchitella — CHANGES_REQUESTED** [review]
Let's close this PR in favour of https://github.com/mendersoftware/deployments/pull/796

**Pull request closed without merging**